### PR TITLE
[RFC] Job control: enable mouse after foregrounding

### PR DESCRIPTION
**Pull request by mhinz**
After foregrounding a neovim process, the terminal emulator's mouse mode gets
enabled according to the 'mouse' option.

References #2335.

### CODE REVIEW

**Comment by Pyrohh** [comment]
Assuming I'm understanding the new UI architecture (it seems really easy to follow), then LGTM.

**Comment by tarruda** [comment]
While this is fine for the current version, I'd like to decouple the TUI from all state kept in the core, which `setmouse` depends on.

@mhinz could you change this PR so that the mouse enable/disable state is kept into the `TUIData` structure? This will be necessary to make it work with remote nvim instances.

**Comment by mhinz** [comment]
@tarruda I split up `setmouse()` into two functions: `setmouse()` and `can_enable_mouse()`. And I added a member `bool can_use_mouse` to `TUIData`. That member is initialized in `tui_start()` via `can_enable_mouse()`.

I hope this is the change you had in mind, but I'm lacking a general overview, so it might be the wrong approach. :-)

**Comment by tarruda** [comment]
@mhinz What I meant is that the process running tui.c can be different than the process running mouse.c, so it shouldn't call core functions like you did. A way fix it would be to expose `can_enable_mouse` via msgpack-rpc, but that would be an overkill for something that is only used by the UI.

The simplest solution is to store the  `tui_mouse_{on,off}` state and restore it from `tui_suspend`, something like this:

```c
static void tui_suspend(UI *ui)
{
  TUIData *data = ui->data;
  bool should_enable_mouse = data->mouse_enabled;
  tui_stop(ui);
  kill(0, SIGTSTP);
  UI* ui = tui_start();
  if (should_enable_mouse) {
    tui_mouse_on(ui);
  } 
}
```

Where `mouse_enabled` is set by `tui_mouse_on`/`tui_mouse_off`. This needs to change `tui_start` to return the new UI instance.

**Comment by mhinz** [comment]
@tarruda Oh, I see. It should work as expected now.

File: src/nvim/tui/tui.c [hunk 1]
```diff
@@ -561,9 +564,14 @@ static void tui_flush(UI *ui)
...
 {
+  TUIData *data = ui->data;
+  bool enable_mouse = data->mouse_enabled;
   tui_stop(ui);
   kill(0, SIGTSTP);
-  tui_start();
+  ui = tui_start();
+  if (enable_mouse) {
+    tui_mouse_on(ui);
+  };
```
Comment (splinterofchaos) [inline src/nvim/tui/tui.c]: superfluous semicolon

Comment (mhinz) [inline src/nvim/tui/tui.c]: *cough*

**Comment by splinterofchaos** [comment]
LGTM :+1:

**Pull request merged**